### PR TITLE
DAOS-8888 misc: Fix ULT leaks

### PR DESCRIPTION
**Pull request by liw**
Argobots dictates the following:

  - A ULT that is not unnamed must be freed explicitly via
    ABT_thread_free. (ABT_thread_join does not do this.)
  - An unnamed ULT cannot be freed explicitly via ABT_thread_free.
  - A ULT cannot free itself.
  - ABT_thread_free implicitly joins the ULT before freeing it.
  - An unnamed ULT cannot be joined via ABT_thread_join.

In several modules, ULTs are only joined but not freed, leading to ULT
leaks. This patch replaces these ABT_thread_join calls with
ABT_thread_free call, which both join as well as free. Also, in some
error paths among those cases, the ULTs may not terminate, leading to
hangs. This patch adopts a check-if-req-is-null-in-the-ULT approach from other
error paths, whose correctness subtly relies on the absence of yielding
between the ULT creation and the ABT_thread_free call---something that
should be replaced with a more robust solution in the future.

sched_req_put does not free the ULT too. This patch adds an "owned" flag
to sched_req so that sched_req_put can call ABT_thread_free on ULTs that
are req-owned. Some dtx ULTs' sched_req_put calls have to be adjusted a
bit to avoid those ULTs from freeing themselves.

Signed-off-by: Li Wei <wei.g.li@intel.com>

### CODE REVIEW

**Review by NiuYawei — DISMISSED** [review]

**Review by Nasf-Fan — DISMISSED** [review]

**Comment by liw** [comment]
The triage decision is to merge this PR and see. (Another diagnostic CI run in #7122 has passed again a few hours ago.)

**Review by jolivier23 — CHANGES_REQUESTED** [review]
This has a conflict with current master and needs a refresh.

File: src/container/srv_container.c [hunk 1]
```diff
@@ -1383,6 +1383,9 @@ cont_agg_eph_leader_ult(void *arg)
 	struct cont_ec_agg	*tmp;
 	int			rc = 0;
 
+	if (svc->cs_ec_leader_ephs_req == NULL)
```
Comment (bfaccini) [inline src/container/srv_container.c:1386]: Why does this become needed ?

File: src/container/srv_target.c [hunk 2]
```diff
@@ -2457,6 +2457,10 @@ ds_cont_tgt_ec_eph_query_ult(void *data)
 
 	D_DEBUG(DB_MD, DF_UUID" start tgt ec query eph ULT\n",
 		DP_UUID(pool->sp_uuid));
+
+	if (pool->sp_ec_ephs_req == NULL)
```
Comment (bfaccini) [inline src/container/srv_target.c:2461]: same question !

**Review by bfaccini — DISMISSED** [review]

File: src/container/srv_container.c (hunk 1 above)
Comment (liw) [reply src/container/srv_container.c:1386]: @bfaccini, these changes correspond to the following in the commit message:

> Also, in some error paths among those cases, the ULTs may not terminate, leading to hangs. This patch adopts a check-if-req-is-null-in-the-ULT approach from other error paths, whose correctness subtly relies on the absence of yielding between the ULT creation and the ABT_thread_free call---something that should be replaced with a more robust solution in the future.

Does this answer your question?

File: src/container/srv_target.c (hunk 2 above)
Comment (liw) [reply src/container/srv_target.c:2461]: Same answer! :)

**Comment by liw** [comment]
Merged with the latest master to fix a minor conflict in src/container/srv_target.c.

**Review by wangdi1 — APPROVED** [review]

**Review by Nasf-Fan — APPROVED** [review]

**Review by NiuYawei — APPROVED** [review]

File: src/container/srv_container.c (hunk 1 above)
Comment (bfaccini) [reply src/container/srv_container.c:1386]: Well, in fact I should have already added that this description was a bit obscure to me ...

Comment (liw) [reply src/container/srv_container.c:1386]: @bfaccini, if cont_svc_ec_agg_leader_start fails to get a sched_req and attempts to handle the error by calling ABT_thread_free to join and free ec_eph_leader_ult, it doesn't do anything to make sure the ULT will actually abort and terminate. Niu and I think the sched_req_get API should be revised to hide this error handling from callers. That is, however, not something we feel comfortable to do for 2.0. This patch therefore adopts the check-if-req-is-null-in-the-ULT approach used by dtx to address the issue.

**Comment by liw** [comment]
F_H_L pool/create_capacity.py: DAOS-8901

File: src/container/srv_container.c (hunk 1 above)
Comment (bfaccini) [reply src/container/srv_container.c:1386]: hmm, ok understood. And I presume this is for the same check-if-req-is-null-in-the-ULT approach that @NiuYawei thinks that to simply fix DAOS-8826/PR-7021 I should just "change sched_req_wakeup() to check passed in 'req' and return if 'req == NULL'" , right ?
But am I wrong to think that this check-if-req-is-null-in-the-ULT approach is a bit racy ??

Comment (liw) [reply src/container/srv_container.c:1386]: @bfaccini, it _is_ racy in the sense that "[its] correctness subtly relies on the absence of yielding between the ULT creation and the ABT_thread_free call", which is true in current code. That's why we feel a more robust solution is required.

**Review by bfaccini — APPROVED** [review]

**Pull request merged**